### PR TITLE
Fix Sparse SGD bug

### PR DESCRIPTION
**Pull request by chengduoZH**
cherry-pick https://github.com/PaddlePaddle/Paddle/pull/13626
test=release/1.0.0

### CODE REVIEW

**Comment by panyx0718** [comment]
what does refine mean? Is it a bug fix or speed up?

File: paddle/fluid/operators/sgd_op.cu [hunk 1]
```diff
@@ -97,13 +96,15 @@ class SGDOpCUDAKernel : public framework::OpKernel<T> {
...
-      dim3 threads(block_size, 1);
-      dim3 grid(1, in_rows.size());
-      SparseSGDFunctorKernel<
-          T, 256><<<grid, threads, 0, ctx.cuda_device_context().stream()>>>(
+      const int kThreadsPerBlock = 256;
+      int thread_x = kThreadsPerBlock;
+      int max_threads = ctx.cuda_device_context().GetMaxPhysicalThreadCount();
+      int max_blocks = std::max(max_threads / kThreadsPerBlock, 1);
+
+      SparseSGDFunctorKernel<<<max_blocks, thread_x, 0,
```
Comment (panyx0718) [inline paddle/fluid/operators/sgd_op.cu:104]: why not use kThreadsPerBlock here?

File: paddle/fluid/operators/sgd_op.cu [hunk 2]
```diff
@@ -97,13 +96,15 @@ class SGDOpCUDAKernel : public framework::OpKernel<T> {
       auto* in_data = in_value.data<T>();
       auto* out_data = param_out->data<T>();
 
-      const int block_size = 256;
-      dim3 threads(block_size, 1);
-      dim3 grid(1, in_rows.size());
```
Comment (panyx0718) [inline paddle/fluid/operators/sgd_op.cu:102]: Is the new way faster?

**Comment by chengduoZH** [comment]
Strictly speaking,this is a bug fix, but only when the input data reaches the limit of Grid size of CUDA, this bug can be triggered. Because the original code sets the grid as the row number of SelecedRows, the code is [here](https://github.com/PaddlePaddle/Paddle/pull/13764/files#diff-aff56a6639f37e774f3d69dd53f6b908L102).

File: paddle/fluid/operators/sgd_op.cu (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/operators/sgd_op.cu:102]: This fix may not make this Op fast, it only makes sure this Op can deal with any size of input.

File: paddle/fluid/operators/sgd_op.cu (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/operators/sgd_op.cu:104]: Of course, here can use kThreadsPerBlock.

**Pull request closed without merging**